syntax = "proto3";

package greet;

option java_package = "com.proto.greet";
option java_multiple_files = true;

import "google/api/annotations.proto";


message Greeting {
  string first_name = 1;
  string last_name = 2;
}

message GreetRequest {
  Greeting greeting = 1;
}

message GreetResponse {
  string result = 1;
}

message GreetManyTimesRequest {
  Greeting greeting = 1;
}

message GreetManyTimesResponse {
  string result = 1;
}

message LongGreetRequest {
  Greeting greeting = 1;
}

message LongGreetResponse {
  string result = 1;
}

message GreetEveryoneRequest {
  Greeting greeting = 1;
}

message GreetEveryoneResponse {
  string result = 1;
}

message GreetWithErrorRequest {
  Greeting greeting = 1;
}

message GreetWithErrorResponse {
  string result = 1;
}

message GreetWithDeadlineRequest {
  Greeting greeting = 1;
}

message GreetWithDeadlineResponse {
  string result = 1;
}

service GreetService {
  // Unary
  rpc Greet(GreetRequest) returns (GreetResponse) {
      option (google.api.http) = {
          get: "/v1/greet/{first_name}"
      };
  }

  // Server streaming
  rpc GreetManyTimes(GreetManyTimesRequest) returns (stream GreetManyTimesResponse) {}

  // Client streaming
  rpc LongGreet(stream LongGreetRequest) returns (LongGreetResponse) {}

  // Bi-Directional streaming
  rpc GreetEveryone(stream GreetEveryoneRequest) returns (stream GreetEveryoneResponse) {}

  // Error handling
  rpc GreetWithError(GreetWithErrorRequest) returns (GreetWithErrorResponse) {}

  // Deadline
  rpc GreetWithDeadline(GreetWithDeadlineRequest) returns (GreetWithDeadlineResponse) {}
}

